cmd/zt_sync_upload_windows_test.go

```diff
@@ -0,0 +1,90 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	chk "gopkg.in/check.v1"
+)
+
+func (s *cmdIntegrationSuite) TestSyncUploadWithExcludeAttrFlag(c *chk.C) {
+	bsu := getBSU()
+
+	srcDirName := scenarioHelper{}.generateLocalDirectory(c)
+	fileList := scenarioHelper{}.generateCommonRemoteScenarioForLocal(c, srcDirName, "")
+
+	// add special files with attributes that we wish to exclude
+	filesToExclude := []string{"file1.pdf", "file2.txt", "file3"}
+	scenarioHelper{}.generateLocalFilesFromList(c, srcDirName, filesToExclude)
+	attrList := []string{"H", "I", "C"}
+	// exclude files with H, I and S attributes
+	excludeAttrsStr := "H;I;S"
+	scenarioHelper{}.setAttributesForLocalFiles(c, srcDirName, filesToExclude, attrList)
+
+	// set up the destination as an empty container
+	containerURL, containerName := createNewContainer(c, bsu)
+	defer deleteContainer(c, containerURL)
+
+	// set up interceptor
+	mockedRPC := interceptor{}
+	Rpc = mockedRPC.intercept
+	mockedRPC.init()
+
+	rawContainerURLWithSAS := scenarioHelper{}.getRawContainerURLWithSAS(c, containerName)
+	raw := getDefaultSyncRawInput(srcDirName, rawContainerURLWithSAS.String())
+	raw.excludeFileAttributes = excludeAttrsStr
+
+	runSyncAndVerify(c, raw, func(err error) {
+		c.Assert(err, chk.IsNil)
+		validateUploadTransfersAreScheduled(c, "", "", fileList, mockedRPC)
+	})
+}
+
+func (s *cmdIntegrationSuite) TestSyncUploadWithIncludeAttrFlag(c *chk.C) {
+	bsu := getBSU()
+
+	srcDirName := scenarioHelper{}.generateLocalDirectory(c)
+	scenarioHelper{}.generateCommonRemoteScenarioForLocal(c, srcDirName, "")
+
+	// add special files with attributes that we wish to include
+	filesToInclude := []string{"file1.txt", "file2.pdf", "file3.pdf"}
+	scenarioHelper{}.generateLocalFilesFromList(c, srcDirName, filesToInclude)
+	includeAttrsStr := "H;I"
+	attrList := []string{"H", "I"}
+	scenarioHelper{}.setAttributesForLocalFiles(c, srcDirName, filesToInclude, attrList)
+
+	// set up the destination as an empty container
+	containerURL, containerName := createNewContainer(c, bsu)
+	defer deleteContainer(c, containerURL)
+
+	// set up interceptor
+	mockedRPC := interceptor{}
+	Rpc = mockedRPC.intercept
+	mockedRPC.init()
+
+	rawContainerURLWithSAS := scenarioHelper{}.getRawContainerURLWithSAS(c, containerName)
+	raw := getDefaultSyncRawInput(srcDirName, rawContainerURLWithSAS.String())
+	raw.includeFileAttributes = includeAttrsStr
+
+	runSyncAndVerify(c, raw, func(err error) {
+		c.Assert(err, chk.IsNil)
+		validateUploadTransfersAreScheduled(c, "", "", filesToInclude, mockedRPC)
+	})
+}
```

Comment:
I wonder if we should also have a test where we are asserting that --include (the name filter) is AND'd with --include-attributes

I.e. use both in the test, and have some files that match only the name filter, some that match only the attribute filter, and some that match both.

That would define the behaviour when combining filters, and alert us if there was ever any accidental change in it.

Reply:
I have a question about these tests. I ran it on my own Windows computer with my own Azure Blob account but the `TestUploadWithIncludeAttrFlag` kept failing because the interceptor obtained more files than expected (number of files created by `generateCommonRemoteScenarioForLocal` + number of files in `filesToInclude`). I then used "H" and "I" flags, which are not default attributes in most Windows files, but it still got more files than what's in the `filesToInclude` list. Any idea why?

Reply:
Hmm. That's weird. Do you know what the extra files were?

Reply:
Not really but they have to come from the 50 files created by `generateCommonRemoteScenarioForLocal`. Last time I ran the test the interceptor caught 33 transfers but only 3 of them were expected, i.e. files from `filesToInclude`. Is there a way to print the list of files from `mockedRPC`?

Reply:
I think the mockedRPC must store the list somewhere (otherwise validateUploadTransfersAreScheduled wouldn't work).  I'd suggest you look at where it stores them, and then temporarily add your own code to just call fmt.PrintLn to dump them out.

Reply:
There was a minor bug, I've fixed it [here](https://github.com/Azure/azure-storage-azcopy/pull/527). We need to use the relative path to construct the full URL, otherwise the get attribute syscall fails, and the filter passes the file, so the all the files inside folders failed, that's where the 30 files came from. I added a INFO output if any error gets returned from the get attribute call, so that it's more visible.